tls/s2n_security_policies.c

```diff
@@ -0,0 +1,554 @@
+            S2N_ERROR_IF(kem_preference->kems!= NULL, S2N_ERR_INVALID_SECURITY_POLICY);
+        }
+    }
+    return 0;
+}
+
+int s2n_ecc_is_extension_required(const struct s2n_security_policy *security_policy)
+{
+    notnull_check(security_policy);
+    for (int i = 0; security_policy_selection[i].version != NULL; i++) {
+        if (security_policy_selection[i].security_policy == security_policy) {
+            return 1 == security_policy_selection[i].ecc_extension_required;
+        }
+    }
+    S2N_ERROR(S2N_ERR_INVALID_SECURITY_POLICY);
+}
+
+int s2n_pq_kem_is_extension_required(const struct s2n_security_policy *security_policy)
+{
+    notnull_check(security_policy);
+    for (int i = 0; security_policy_selection[i].version != NULL; i++) {
+        if (security_policy_selection[i].security_policy == security_policy) {
+            return 1 == security_policy_selection[i].pq_kem_extension_required;
+        }
+    }
+    S2N_ERROR(S2N_ERR_INVALID_SECURITY_POLICY);
+}
+
+/* Checks whether cipher preference supports TLS 1.3 based on whether it is configured
+ * with TLS 1.3 ciphers. Returns true or false.
+ */
+bool s2n_security_policy_supports_tls13(const struct s2n_security_policy *security_policy)
+{
+    if (security_policy == NULL) {
+        return false;
+    }
+
+    for (uint8_t i = 0; security_policy_selection[i].version != NULL; i++) {
+        if (security_policy_selection[i].security_policy == security_policy) {
+            return security_policy_selection[i].supports_tls13 == 1;
+        }
+    }
+
+    /* if cipher preference is not in the official list, compute the result */
+    const struct s2n_cipher_preferences *cipher_preferences = security_policy->cipher_preferences;
+    notnull_check(cipher_preferences);
+    for (uint8_t i = 0; i < cipher_preferences->count; i++) {
+        if (s2n_is_valid_tls13_cipher(cipher_preferences->suites[i]->iana_value)) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+int s2n_connection_is_valid_for_cipher_preferences(struct s2n_connection *conn, const char *version)
+{
+    notnull_check(conn);
+    notnull_check(version);
+    notnull_check(conn->secure.cipher_suite);
+
+    const struct s2n_security_policy *security_policy = NULL;
+    GUARD(s2n_find_security_policy_from_version(version, &security_policy));
+    notnull_check(security_policy);
+
+    /* make sure we dont use a tls version lower than that configured by the version */
+    if (s2n_connection_get_actual_protocol_version(conn) < security_policy->minimum_protocol_version) {
+        return 0;
+    }
+
+    struct s2n_cipher_suite *cipher = conn->secure.cipher_suite;
+    notnull_check(cipher);
+    for (int i = 0; i < security_policy->cipher_preferences->count; ++i) {
+        if (0 == memcmp(security_policy->cipher_preferences->suites[i]->iana_value, cipher->iana_value, S2N_TLS_CIPHER_SUITE_LEN)) {
+            return 1;
```

Comment:
Boolean issue here, if using `GUARD`s and `notnull_check` et al, we should always return 0 on success. Maybe this should be a boolean, or should take an out parameter that indicates success?

Reply:
will return 0 on success